src/orchestrator-kubernetes/src/lib.rs

```diff
@@ -437,35 +519,209 @@ fn parse_k8s_quantity(s: &str) -> Result<ScaledQuantity, anyhow::Error> {
+                        // too since we only use disk for spilling memory.
+                        let exit_code = termination_state.map(|s| s.exit_code);
+                        exit_code == Some(135) || exit_code == Some(137)
+                    })
+                })
+                .unwrap_or(false);
+
+            let (pod_ready, last_probe_time) = pod
+                .status
+                .and_then(|status| status.conditions)
+                .and_then(|conditions| conditions.into_iter().find(|c| c.type_ == "Ready"))
+                .map(|c| (c.status == "True", c.last_probe_time))
+                .unwrap_or((false, None));
+
+            let status = if pod_ready {
+                ServiceStatus::Ready
+            } else {
+                ServiceStatus::NotReady(oomed.then_some(NotReadyReason::OomKilled))
+            };
+            let time = if let Some(time) = last_probe_time {
+                time.0
+            } else {
+                Utc::now()
+            };
+
+            Ok(ServiceEvent {
+                service_id,
+                process_id,
+                status,
+                time,
+            })
+        }
+
+        let stream = watcher(self.config.pod_api.clone(), self.watch_pod_params())
+            .touched_objects()
+            .filter_map(|object| async move {
+                match object {
+                    Ok(pod) => Some(into_service_event(pod)),
+                    Err(error) => {
+                        // We assume that errors returned by Kubernetes are usually transient, so we
+                        // just log a warning and ignore them otherwise.
+                        tracing::warn!("service watch error: {error}");
+                        None
+                    }
+                }
+            });
+        Box::pin(stream)
+    }
+
+    async fn fetch_service_metrics(
+        &self,
+        id: &str,
+    ) -> Result<Vec<ServiceProcessMetrics>, anyhow::Error> {
+        let (result_tx, result_rx) = oneshot::channel();
+        self.send_command(OrchestratorCommand::FetchServiceMetrics {
+            id: id.to_string(),
+            result_tx,
+        });
+
+        result_rx.await.expect("task not dropped")
+    }
+
+    fn update_scheduling_config(&self, config: ServiceSchedulingConfig) {
+        self.send_command(OrchestratorCommand::UpdateSchedulingConfig { config });
+    }
+}
+
+impl NamespacedKubernetesOrchestratorTask {
+    fn spawn(self) -> AbortOnDropHandle<()> {
+        let name = format!("kubernetes-orchestrator:{}", self.config.namespace);
+        mz_ore::task::spawn(|| name, self.run()).abort_on_drop()
+    }
+
+    async fn run(mut self) {
+        while let Some(cmd) = self.command_rx.recv().await {
+            use OrchestratorCommand::*;
+            let result = match cmd {
+                EnsureService { id, config } => self.ensure_service(id, config).await,
+                DropService { id } => self.drop_service(&id).await,
```

Comment:
I think we want to retry infinitely if either of these fails, rather than taking down envd. (Should also send an `error!` to Sentry because we need to immediately manually remediate.)